### PR TITLE
HIVE-25990: Optimise multiple copies in case of CTAS in external tables for Object stores.

### PR DESCRIPTION
**Pull request by ayushtkn**

### CODE REVIEW

**Comment by ayushtkn** [comment]
the test failure isn't related.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 1]
```diff
@@ -1514,6 +1523,18 @@ public static void mvFileToFinalPath(Path specPath, Configuration hconf,
...
   }
 
+  private static void createFileList(Set<FileStatus> filesKept, Path srcPath, Path targetPath, FileSystem fs)
+      throws IOException {
+    String files = srcPath.toString() + System.lineSeparator();
+    for (FileStatus file : filesKept) {
+      files += file.getPath().toString() + System.lineSeparator();
+    }
+    try (FSDataOutputStream outStream = fs.create(new Path(targetPath, BLOB_FILES_KEPT))) {
+      outStream.writeBytes(files);
```
Comment (rbalamohan) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:1533]: Why not loop through filesKept here and directly write them to outStream? That will mem pressure in constructing "files"

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 2]
```diff
@@ -1514,6 +1523,17 @@ public static void mvFileToFinalPath(Path specPath, Configuration hconf,
     fs.delete(taskTmpPath, true);
   }
 
+  private static void createFileList(Set<FileStatus> filesKept, Path srcPath, Path targetPath, FileSystem fs)
+      throws IOException {
+    try (FSDataOutputStream outStream = fs.create(new Path(targetPath, BLOB_FILES_KEPT))) {
+      outStream.writeBytes(srcPath.toString() + System.lineSeparator());
```
Comment (rbalamohan) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:1529]: Is it needed to write the srcPath as first entry?  Only file paths can be written out and during reading, first entry's parent directory can be assumed as srcPath?

Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:1531]: Have added a comment above explaining the reason, as discussed offline

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 3]
```diff
@@ -1496,7 +1499,13 @@ public static void mvFileToFinalPath(Path specPath, Configuration hconf,
           // for CTAS or Create MV statements
           perfLogger.perfLogBegin("FileSinkOperator", "moveSpecifiedFileStatus");
           LOG.debug("CTAS/Create MV: Files being renamed:  " + filesKept.toString());
-          moveSpecifiedFilesInParallel(hconf, fs, tmpPath, specPath, filesKept);
+          if (conf.getTable() != null && conf.getTable().getTableType().equals(TableType.EXTERNAL_TABLE)) {
+            // Do this optimisation only for External tables.
+            createFileList(filesKept, tmpPath, specPath, fs);
+          } else {
+            List<String> filesKeptPaths = filesKept.stream().map(x -> x.getPath().toString()).collect(Collectors.toList());
```
Comment (rbalamohan) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:1506]: filesKept is a set. Is it possible to retain as "set" after the mapping? (Something like Collectors.toSet()). if so, no need to change the signature of moveSpecifiedFilesInParallel?

Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java]: Done, Changed to Set.

**Comment by rbalamohan** [comment]
LGTM. +1.

**Comment by ayushtkn** [comment]
Test failure not related. Failing in other PRs as well

**Pull request merged**